storage/paths/db_index.go

```diff
@@ -0,0 +1,891 @@
+	}
+
+	var rows []struct {
+		StorageId  string
+		Urls       string
+		Weight     uint64
+		MaxStorage uint64
+		CanSeal    bool
+		CanStore   bool
+		Groups     string
+		AllowTo    string
+		AllowTypes string
+		DenyTypes  string
+	}
+
+	sql := fmt.Sprintf(`SELECT storage_id, 
+								urls, 
+								weight,
+								max_storage, 
+								can_seal, 
+								can_store, 
+								groups, 
+								allow_to, 
+								allow_types, 
+								deny_types 
+						 FROM storagelocation 
+						 WHERE %s and available >= $1
+						 and NOW()-last_heartbeat < $2 
+						 and heartbeat_err is null
+						order by (available::numeric * weight) desc`, checkString)
+	err = dbi.harmonyDB.Select(ctx, &rows,
+		sql, spaceReq, SkippedHeartbeatThresh)
+	if err != nil {
+		return nil, xerrors.Errorf("Querying for best storage sectors fails with sql: %s and err %w: ", sql, err)
+	}
+
+	var result []storiface.StorageInfo
+	for _, row := range rows {
+		result = append(result, storiface.StorageInfo{
+			ID:         storiface.ID(row.StorageId),
+			URLs:       splitString(row.Urls),
+			Weight:     row.Weight,
+			MaxStorage: row.MaxStorage,
+			CanSeal:    row.CanSeal,
+			CanStore:   row.CanStore,
+			Groups:     splitString(row.Groups),
+			AllowTo:    splitString(row.AllowTo),
+			AllowTypes: splitString(row.AllowTypes),
+			DenyTypes:  splitString(row.DenyTypes),
+		})
+	}
+
+	return result, nil
+}
+
+// timeout after which we consider a lock to be stale
+const LockTimeOut = 300 * time.Second
+
+func (dbi *DBIndex) lock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) (bool, error) {
+	if read|write == 0 {
+		return false, nil
+	}
+
+	if read|write > (1<<storiface.FileTypes)-1 {
+		return false, xerrors.Errorf("unknown file types specified")
+	}
+
+	var rows []struct {
+		SectorFileType storiface.SectorFileType
+		ReadTs         sql.NullTime
+		ReadRefs       int
+		WriteTs        sql.NullTime
+	}
+	_, err := dbi.harmonyDB.BeginTransaction(ctx, func(tx *harmonydb.Tx) (commit bool, err error) {
+
+		fts := (read | write).AllSet()
+		err = dbi.harmonyDB.Select(ctx, &rows,
+			`				SELECT sector_filetype, read_ts, read_refs, write_ts 
+								 FROM sectorlocation 
+								 WHERE miner_id=$1 
+								   AND sector_num=$2
+								   AND sector_filetype = ANY($3)`,
+			sector.Miner, sector.Number, fts)
+		if err != nil {
+			return false, xerrors.Errorf("StorageLock SELECT fails: %v", err)
+		}
+
+		type locks struct {
+			readTs   sql.NullTime
+			readRefs int
+			writeTs  sql.NullTime
+		}
+		lockMap := make(map[storiface.SectorFileType]locks)
+		for _, row := range rows {
+			lockMap[row.SectorFileType] = locks{
+				readTs:   row.ReadTs,
+				readRefs: row.ReadRefs,
+				writeTs:  row.WriteTs,
+			}
+		}
+
+		// Check if we can acquire write locks
+		// Conditions: No write lock or write lock is stale, No read lock or read lock is stale
+		for wft := range write.AllSet() {
```

Comment:
Pretty sure this should be
```suggestion
		for _, wft := range write.AllSet() {
```